### ISSUE TITLE
Logical error: 'Broken buckets. its size(5) is not power of 2'.

### ISSUE THREAD

**Issue by qoega** [ClickHouse/ClickHouse#51703]
https://s3.amazonaws.com/clickhouse-test-reports/0/a8e0dbf1d32b23cb4f6a81dbcf6dede35316d183/stress_test__tsan_.html

```
 [ 23523 ] {1287f52a-b279-4ca0-ac8f-6770b18db8ca} <Fatal> : Logical error: \'Broken buckets. its size(5) is not power of 2\'.
 [ 27219 ] {} <Fatal> BaseDaemon: ########################################
 [ 27219 ] {} <Fatal> BaseDaemon: (version 23.7.1.1 (official build), build id: C4C042E1BA38FC371C46B8D905A87E5EBA49DCCB, git hash: a8e0dbf1d32b23cb4f6a81dbcf6dede35316d183) (from thread 23523) (query_id: 1287f52a-b279-4ca0-ac8f-6770b18db8ca) (query: SELECT t1.key, t2.key, t1.s, empty(t2.s) FROM t1
 [ 27219 ] {} <Fatal> BaseDaemon: 
 [ 27219 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007f3c43352a7c 0x00007f3c432fe476 0x00007f3c432e47f3 0x000055d6fb57f8a3 0x000055d70394f689 0x000055d70394ff60 0x000055d6fd3b0c6d 0x000055d70cb0c32d 0x000055d70cb0a829 0x000055d70efc458d 0x000055d70ec7c472 0x000055d70ec6f27f 0x000055d70ec70271 0x0000
 [ 27219 ] {} <Fatal> BaseDaemon: 5. pthread_kill @ 0x00007f3c43352a7c in ?
 [ 27219 ] {} <Fatal> BaseDaemon: 6. raise @ 0x00007f3c432fe476 in ?
 [ 27219 ] {} <Fatal> BaseDaemon: 7. abort @ 0x00007f3c432e47f3 in ?
 [ 27219 ] {} <Fatal> BaseDaemon: 8. abort @ 0x000000000dcd48a3 in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 9. ./build_docker/./src/Common/Exception.cpp:0: DB::abortOnFailedAssertion(String const&) @ 0x00000000160a4689 in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 10.1. inlined from ./build_docker/./src/Common/Exception.cpp:66: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&)
 [ 27219 ] {} <Fatal> BaseDaemon: 10. ./build_docker/./src/Common/Exception.cpp:96: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x00000000160a4f60 in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 11. DB::Exception::Exception<unsigned long>(int, FormatStringHelperImpl<std::type_identity<unsigned long>::type>, unsigned long&&) @ 0x000000000fb05c6d in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 12. ./build_docker/./src/Interpreters/GraceHashJoin.cpp:644: DB::GraceHashJoin::addJoinedBlockImpl(DB::Block) @ 0x000000001f26132d in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 13. ./build_docker/./src/Interpreters/GraceHashJoin.cpp:0: DB::GraceHashJoin::addJoinedBlock(DB::Block const&, bool) @ 0x000000001f25f829 in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 14. ./build_docker/./src/Processors/Transforms/JoiningTransform.cpp:308: DB::FillingRightJoinSideTransform::work() @ 0x000000002171958d in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 15.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/list:588: std::__list_imp<DB::ExecutingGraph::Edge, std::allocator<DB::ExecutingGraph::Edge>>::__sz[abi:v15000]() const
 [ 27219 ] {} <Fatal> BaseDaemon: 15.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/list:616: std::__list_imp<DB::ExecutingGraph::Edge, std::allocator<DB::ExecutingGraph::Edge>>::empty[abi:v15000]() const
 [ 27219 ] {} <Fatal> BaseDaemon: 15.3. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/list:918: std::list<DB::ExecutingGraph::Edge, std::allocator<DB::ExecutingGraph::Edge>>::empty[abi:v15000]() const
 [ 27219 ] {} <Fatal> BaseDaemon: 15.4. inlined from ./build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:50: DB::executeJob(DB::ExecutingGraph::Node*, DB::ReadProgressCallback*)
 [ 27219 ] {} <Fatal> BaseDaemon: 15. ./build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:92: DB::ExecutionThreadContext::executeTask() @ 0x00000000213d1472 in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 16. ./build_docker/./src/Processors/Executors/PipelineExecutor.cpp:255: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x00000000213c427f in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 17.1. inlined from ./build_docker/./src/Processors/Executors/PipelineExecutor.cpp:351: operator()
 [ 27219 ] {} <Fatal> BaseDaemon: 17.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394: decltype(std::declval<DB::PipelineExecutor::spawnThreads()::$_0&>()()) std::__invoke[abi:v15000]<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spaw
 [ 27219 ] {} <Fatal> BaseDaemon: 17.3. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479: void std::__invoke_void_return_wrapper<void, true>::__call<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&)
 [ 27219 ] {} <Fatal> BaseDaemon: 17.4. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235: std::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>::operator()[abi:v15000]()
 [ 27219 ] {} <Fatal> BaseDaemon: 17. ./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>>(std::__function::__policy_st
 [ 27219 ] {} <Fatal> BaseDaemon: 18.1. inlined from ./build_docker/./base/base/../base/wide_integer_impl.h:809: bool wide::integer<128ul, unsigned int>::_impl::operator_eq<wide::integer<128ul, unsigned int>>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
 [ 27219 ] {} <Fatal> BaseDaemon: 18.2. inlined from ./build_docker/./base/base/../base/wide_integer_impl.h:1482: bool wide::operator==<128ul, unsigned int, 128ul, unsigned int>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
 [ 27219 ] {} <Fatal> BaseDaemon: 18.3. inlined from ./build_docker/./base/base/../base/strong_typedef.h:42: StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>::operator==(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&) const
 [ 27219 ] {} <Fatal> BaseDaemon: 18.4. inlined from ./build_docker/./src/Common/OpenTelemetryTraceContext.h:65: DB::OpenTelemetry::Span::isTraceEnabled() const
 [ 27219 ] {} <Fatal> BaseDaemon: 18. ./build_docker/./src/Common/ThreadPool.cpp:429: ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x00000000161cccd2 in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 19. ./build_docker/./src/Common/ThreadPool.cpp:0: ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::\'lambda0\'()>(v
 [ 27219 ] {} <Fatal> BaseDaemon: 20. ./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:717: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<Thr
 [ 27219 ] {} <Fatal> BaseDaemon: 21.1. inlined from ./build_docker/./base/base/../base/wide_integer_impl.h:809: bool wide::integer<128ul, unsigned int>::_impl::operator_eq<wide::integer<128ul, unsigned int>>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
 [ 27219 ] {} <Fatal> BaseDaemon: 21.2. inlined from ./build_docker/./base/base/../base/wide_integer_impl.h:1482: bool wide::operator==<128ul, unsigned int, 128ul, unsigned int>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
 [ 27219 ] {} <Fatal> BaseDaemon: 21.3. inlined from ./build_docker/./base/base/../base/strong_typedef.h:42: StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>::operator==(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&) const
 [ 27219 ] {} <Fatal> BaseDaemon: 21.4. inlined from ./build_docker/./src/Common/OpenTelemetryTraceContext.h:65: DB::OpenTelemetry::Span::isTraceEnabled() const
 [ 27219 ] {} <Fatal> BaseDaemon: 21. ./build_docker/./src/Common/ThreadPool.cpp:429: ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x00000000161c89c7 in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 22.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>::reset[abi:v15000](std::__thread_struct*)
 [ 27219 ] {} <Fatal> BaseDaemon: 22.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
 [ 27219 ] {} <Fatal> BaseDaemon: 22.3. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/tuple:265: ~__tuple_leaf
 [ 27219 ] {} <Fatal> BaseDaemon: 22.4. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/tuple:538: ~tuple
 [ 27219 ] {} <Fatal> BaseDaemon: 22.5. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: std::default_delete<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(st
 [ 27219 ] {} <Fatal> BaseDaemon: 22.6. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:305: std::unique_ptr<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::
 [ 27219 ] {} <Fatal> BaseDaemon: 22.7. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
 [ 27219 ] {} <Fatal> BaseDaemon: 22. ./build_docker/./contrib/llvm-project/libcxx/include/thread:297: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function
 [ 27219 ] {} <Fatal> BaseDaemon: 23. __tsan_thread_start_func @ 0x000000000dccddef in /usr/bin/clickhouse
 [ 27219 ] {} <Fatal> BaseDaemon: 24. ? @ 0x00007f3c43350b43 in ?
 [ 27219 ] {} <Fatal> BaseDaemon: 25. ? @ 0x00007f3c433e2a00 in ?
 [ 27219 ] {} <Fatal> BaseDaemon: Integrity check of the executable successfully passed (checksum: CA682D1A7B965E8377C92CB5C2D42A27)
 [ 1713 ] {} <Fatal> Application: Child process was terminated by signal 6.
```

**Comment by qoega** [comment]
cc: @vdimir

**Comment by qoega** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/0/3e8358e8a1be3e8efe525584df99cd50f7f8fc1c/stress_test__asan_.html

**Comment by devcrafter** [comment]
Query and settings from stress test logs - https://pastila.nl/?034d5871/ee6879766b3cf3518cdd5c60f425f30a

**Comment by zhanglistar** [comment]
This is not "logic error", it normal exception, it occurs when new file can't be created, say reaching OS inode or capacity limit, or OOM when creating new file like the stress test case, then we have not correct buckets. So this should be a normal exception not logic error. Maybe later we will submit a PR to modify related code.

**Comment by devcrafter** [comment]
> This is not "logic error", it normal exception, it occurs when new file can't be created, say reaching OS inode or capacity limit, or OOM when creating new file like the stress test case, then we have not correct buckets. So this should be a normal exception not logic error. Maybe later we will submit a PR to modify related code to make sure that bucket rehashing process is atomic to make things straightforward.

@zhanglistar Thanks for clarification. Agree. Still, an additional check wouldn't hurt, but in other place. See Done in [30be0ab](https://github.com/ClickHouse/ClickHouse/pull/51737/commits/30be0ab4a8af7247a12c94076cd17834f712b9d9)

**Comment by antonio2368** [comment]
Seems like it's still happening https://s3.amazonaws.com/clickhouse-test-reports/52373/5decb1f5c555d2465724f9bc3c555c157f9deb81/upgrade_check__asan_.html

**Comment by devcrafter** [comment]
> Seems like it's still happening https://s3.amazonaws.com/clickhouse-test-reports/52373/5decb1f5c555d2465724f9bc3c555c157f9deb81/upgrade_check__asan_.html

It's happening to version `23.6.2.18`, the version we upgraded from (`run.log`):  
```
Unpacking clickhouse-server (23.7.1.1+asan) over (23.6.2.18) ...
```

**Comment by vdimir** [comment]
Recently I disabled test from upgrade check not to face this error https://github.com/ClickHouse/ClickHouse/pull/52235

Should we disable more tests that reproduces this error until fix is in release version?

**Comment by devcrafter** [comment]
> Recently I disabled test from upgrade check not to face this error #52235
> 
> Should we disable more tests that reproduces this error until fix is in release version?

Wouldn't it be better to back port the fix? I'm not sure if Upgrade check will pick up the new version

**Comment by tavplubix** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/52558/2efbeab5afe50fbd734a6729e4cffa7ef12fff04/upgrade_check__debug_.html

**Comment by devcrafter** [comment]
> https://s3.amazonaws.com/clickhouse-test-reports/52558/2efbeab5afe50fbd734a6729e4cffa7ef12fff04/upgrade_check__debug_.html

It's the same as https://github.com/ClickHouse/ClickHouse/issues/51703#issuecomment-1645414000

@tavplubix If we back port the fix to 23.6, - will the upgrade check use the new 23.6 version?

**Comment by tavplubix** [comment]
> It's the same as https://github.com/ClickHouse/ClickHouse/issues/51703#issuecomment-1645414000

Yes, and we should fix it.

> If we back port the fix to 23.6, - will the upgrade check use the new 23.6 version?

Yes, after releasing a new patch-version of 23.6
